### PR TITLE
refactor!: split `r/demo/users`

### PR DESCRIPTION
**Pull request by harry-hov**
Splits `r/demo/users` into `p/demo/users` and `r/demo/users` 

Related #1393 

<details><summary>Contributors' checklist...</summary>

- [x] Added new tests, or not needed, or not feasible
- [x] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [x] Updated the official documentation or not needed
- [x] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [x] Added references to related issues and PRs
- [x] Provided any useful hints for running manual tests
- [x] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

File: examples/gno.land/p/demo/grc/grc1155/gno.mod
Comment (moul) [inline examples/gno.land/p/demo/grc/grc1155/gno.mod:6]: p/ should not require r/

File: examples/gno.land/p/demo/users/users.gno
Comment (deelawn) [inline examples/gno.land/p/demo/users/users.gno:20]: This might be a good time to rename this method to `String()` so it gets the advantage of implementing the `fmt.Stringer` interface for when we eventually support it. I do realize it was already named `Render` and you just moved it; I'm just not sure the name is helpful since `Render` is usually used to mark a function that produces text to be served via gnoweb.

File: examples/gno.land/p/demo/users/users.gno
Comment (deelawn) [inline examples/gno.land/p/demo/users/users.gno:14]: Are fields like profile, number, invites, and inviter needed as part of a basic `User` definition?  These seem to be application specific.  Or at the very least I should be able to write an application that can couple a user's address with its username without the extra fields I don't care about; that seems to be the base functionality.

But it also might be helpful to provide a struct like `UserData` or `Extended` user that provides the extra fields for like applications that want to operate using this data model:
```
type ExtendedUser struct {
    User
    Profile string
    Number int
    Invites int
    Inviter std.Address
}
```

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token_test.gno
Comment (deelawn) [inline examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token_test.gno:38]: Is it necessary to still include the `users.AddressOrName` cast if it isn't calling `Resolve`?  There are lots of other places that still do this as well.

File: examples/gno.land/p/demo/users/types.gno
Comment (deelawn) [inline examples/gno.land/p/demo/users/types.gno:5]: These two methods also seem very application specific.  I don't think there is any obvious reason to prefix usernames with `@` in general.  Though I do think this type can still be helpful to differentiate a string between being an address or name.  This might be a good opportunity to make each method accept a `usernamePrefix string` argument that allows the application to decide this.

File: examples/gno.land/r/demo/users/users.gno
Comment (deelawn) [inline examples/gno.land/r/demo/users/users.gno:18]: A user storage structure seems like a great candidate for something that should be included in a users package.  It makes it easy for applications to maintain and access their own user list.  Something like:
```
type p/users.Storage struct {
    nameToUser *avl.Tree
    addresstoUser *avl.Tree
}

func (s *Storage) GetUserByName(name string) *User
func (s *Storage) GetExtendeUserByName(name string) *ExtendedUser
func (s *Storage) getUserByName(name string) interface{}{ return s.nameToUser.Get(name) }

etc
...
```

**Review by deelawn — COMMENTED** [review]
I think this is a great idea.  But I also think this can be an opportunity to take this a bit further and separate additional application logic from what we might expect to be a part of a general package any realm can use to manage users.

File: examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token_test.gno
Comment (harry-hov) [reply examples/gno.land/p/demo/grc/grc1155/basic_grc1155_token_test.gno]: Fixed: 9ee009bffb21fb419c17b424719f6c1c98f946d3

File: examples/gno.land/p/demo/users/users.gno
Comment (harry-hov) [reply examples/gno.land/p/demo/users/users.gno]: Done 8ea3f371a2be2b606a9840c6c287467159ae3de2

**Comment by harry-hov** [comment]
@deelawn I love your suggestions 💯  However, I would like to stick to refactoring for this PR. I prefer not to introduce any changes to the logic. The main motivation behind this PR is to unblock PR #1393. Once this PR is merged, I will sum up your opinions in a separate issue. WDYT?

**Comment by deelawn** [comment]
> @deelawn I love your suggestions 💯 However, I would like to stick to refactoring for this PR. I prefer not to introduce any changes to the logic. The main motivation behind this PR is to unblock PR #1393. Once this PR is merged, I will sum up your opinions in a separate issue. WDYT?

Yeah fair enough 👍

File: examples/gno.land/p/demo/microblog/microblog.gno
Comment (thehowl) [inline examples/gno.land/p/demo/microblog/microblog.gno:1]: Hmm, I think we should discuss this for a second. I think what you did makes sense, but it's still worth agreeing on the best way to answer the following:

**Should render functions sit within a p/ package or exclusively within r/?**

Note we can have the render functions keep sitting here. We could have a `type UserStore interface` declared in `p/demo/users`, which has an implementation in `r/demo/users`, and the `UserStore` is placed within Microblog. This would enable `p/` packages to keep "plugging in" to users, while having to make the connection explicit in the code of a realm. (This may be a good approach regardless).

But, we could also simply say "rendering should be decided directly by the realm". In which case, what you did here makes sense.

I have no strong preference either way, I do think p/ should implement "business logic" while it makes sense for r/ to be more like "apps" (and as such more concerned with rendering, decoupled from the business logic itself). In any case, I think whatever we do, we need to be consistent.

File: examples/gno.land/p/demo/users/types.gno
Comment (thehowl) [reply examples/gno.land/p/demo/users/types.gno:5]: @deelawn I think the way it currently is makes a lot of sense!

r/demo/users is, to me, the "username" realm every other realm should use to identify users. p/demo/users is not meant to be the "universal" implementation for all user systems.

If we have a realm function which is, say, the following:

```go
func SendMessage(dst users.AddressOrName, title, content string)
```

having users.AddressOrName as the type indicates that, unless the realm's doing something weird, the function will accept a full address or also a username prefixed with `@`. So I'm personally in favour to keep it this way; `p/demo/users` is allowed to have some opinions :)

File: examples/gno.land/p/demo/users/users.gno
Comment (thehowl) [inline examples/gno.land/p/demo/users/users.gno:20]: with respect to my first comment, I think it makes sense to consistently place this in r/demo/users if we decide to keep rendering within realms.

also, I think that if we do keep it here, this should be `Render() string` (as it was originally). Another idea for a "rendering function" --  `Link() string` -- which returns `[@user](/r/demo/users:user)` -- this can be deferred to another PR.

Since `String() string` is mostly useful for debugging (as formatters understand String() very often), we could have `String()` return something like `@user (<address>)`.

File: examples/gno.land/p/demo/users/users.gno
Comment (thehowl) [reply examples/gno.land/p/demo/users/users.gno:14]: @deelawn agreed, I think it makes sense to keep those fields relevant to r/demo/users (when rendering pages and using its inviting system) so that third party realms/packages only get what they need most of the time (a name associated to an address)

**Review by thehowl — COMMENTED** [review]
I think there's a "guideline" decision to be made here and to place on Effective Gno. Aside from that, the split looks good so pre-approving pending deciding what to do. (Please, in the meantime, do change the method where I commented from String to Render, as it was originally)

**Review by thehowl — APPROVED** [review]

File: examples/gno.land/p/demo/users/users.gno
Comment (harry-hov) [reply examples/gno.land/p/demo/users/users.gno]: Done: c894748e83fd132a51eac1d3d7ba041b3badd36d

File: examples/gno.land/p/demo/microblog/microblog.gno
Comment (thehowl) [reply examples/gno.land/p/demo/microblog/microblog.gno:1]: cc/ @moul 

read the above comment -- I raised the question in the review meeting and there's no strong preference; a slight inclination towards preferring rendering in r/ (me & leon), hariom thinks that there should be no "guideline" set in this case. Tagging you as I think you may have some argument either way; but going ahead and merging this for now (and can apply any further chagnes later)

Comment (moul) [reply examples/gno.land/p/demo/microblog/microblog.gno:1]: I also do that for `r/demo/blog` (https://github.com/gnolang/gno/blob/master/examples/gno.land/r/gnoland/blog/gnoblog.gno#L23-L25). It makes sense to declare the `Render` function in helpers. But instead of naming it `Render()`, consider `RenderBlock`, `RenderStats`, `RenderSomething`, or `func (o *obj) Render()`. Avoid having top-level Render functions in `p/`. This could be a good warning for a linter.

In my opinion, as you mentioned, `p/` should not handle `business logic`. However, it can define `business logic helpers`.

**Comment by jaekwon** [comment]
Thank you Hariom!

**Pull request merged**
